pkg/util/downloader_test.go

```diff
@@ -0,0 +1,121 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"k8s.io/minikube/pkg/minikube/constants"
+	"k8s.io/minikube/pkg/minikube/tests"
+)
+
+func TestGetISOFileURI(t *testing.T) {
+	dler := DefaultDownloader{}
+
+	tests := map[string]string{
+		"file:///test/path/minikube-test.iso":                           "file:///test/path/minikube-test.iso",
+		"https://storage.googleapis.com/minikube/iso/minikube-test.iso": "file://" + filepath.ToSlash(filepath.Join(constants.GetMinipath(), "cache", "iso", "minikube-test.iso")),
+	}
+
+	for input, expected := range tests {
+		if isoFileURI := dler.GetISOFileURI(input); isoFileURI != expected {
+			t.Fatalf("Expected GetISOFileURI with input %s to return %s but instead got: %s", input, expected, isoFileURI)
+		}
+	}
+
+}
+
+var testISOString = "hello"
+
+type ISOHandlerCorrect struct{}
+
+func (h *ISOHandlerCorrect) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	io.WriteString(w, testISOString)
+}
+
+type ISOHandlerError struct{}
+
+func (h *ISOHandlerError) ServeHTTP(w http.ResponseWriter, r *http.Request) {}
+
+func TestCacheMinikubeISOFromURL(t *testing.T) {
+	tempDir := tests.MakeTempDir()
+	defer os.RemoveAll(tempDir)
+	dler := DefaultDownloader{}
+	isoPath := filepath.Join(constants.GetMinipath(), "cache", "iso", "minikube-test.iso")
+
+	server := httptest.NewServer(&ISOHandlerCorrect{})
+	isoURL := server.URL + "/minikube-test.iso"
+	if err := dler.CacheMinikubeISOFromURL(isoURL); err != nil {
+		t.Fatalf("Unexpected error from CacheMinikubeISOFromURL: %s", err)
+	}
+
+	transferred, err := ioutil.ReadFile(filepath.Join(isoPath))
+	if err != nil {
+		t.Fatalf("File not copied. Could not open file at path: %s", isoPath)
+	}
+
+	//test that the ISO is transferred properly
+	contents := []byte(testISOString)
+	if !bytes.Contains(transferred, contents) {
+		t.Fatalf("Expected transfers to contain: %s. It was: %s", contents, transferred)
+	}
+
+}
+
+func TestShouldCacheMinikubeISO(t *testing.T) {
+	dler := DefaultDownloader{}
+
+	tests := map[string]bool{
+		"file:///test/path/minikube-test.iso":                           false,
+		"https://storage.googleapis.com/minikube/iso/minikube-test.iso": true,
+	}
+
+	for input, expected := range tests {
+		if out := dler.ShouldCacheMinikubeISO(input); out != expected {
+			t.Fatalf("Expected ShouldCacheMinikubeISO with input %s to return %d but instead got: %t", input, expected, out)
+		}
+	}
+}
+
+func TestIsMinikubeISOCached(t *testing.T) {
+	tempDir := tests.MakeTempDir()
```

Comment:
tempDir is for the ioutil.Writefil, this way the file is written in /tmp. tempDir modifies constants.GetMinipath() which is used.  tempDir is a var so that it can be closed w/ defer